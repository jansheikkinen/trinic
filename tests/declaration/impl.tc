// Impl

struct TestStruct
  x, y: int8
end

trait TestTrait
  function testFunction(self: *Self) -> Self
end

impl TestTrait for TestStruct
  function testFunction(self: *Self) -> Self do
    return(self);
  end
end
