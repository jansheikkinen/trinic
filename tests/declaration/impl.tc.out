
### LEXER ###
000:       STRUCT,     (null), (003, 006)
001:   IDENTIFIER, TestStruct, (003, 017)
002:   IDENTIFIER,          x, (004, 003)
003:            ,,     (null), (004, 004)
004:   IDENTIFIER,          y, (004, 006)
005:            :,     (null), (004, 007)
006:         INT8,     (null), (004, 012)
007:          END,     (null), (005, 003)
008:        TRAIT,     (null), (007, 005)
009:   IDENTIFIER,  TestTrait, (007, 015)
010:     FUNCTION,     (null), (008, 010)
011:   IDENTIFIER, testFunction, (008, 023)
012:            (,     (null), (008, 024)
013:   IDENTIFIER,       self, (008, 028)
014:            :,     (null), (008, 029)
015:            *,     (null), (008, 031)
016:   IDENTIFIER,       Self, (008, 035)
017:            ),     (null), (008, 036)
018:           ->,     (null), (008, 038)
019:   IDENTIFIER,       Self, (008, 043)
020:          END,     (null), (009, 003)
021:         IMPL,     (null), (011, 004)
022:   IDENTIFIER,  TestTrait, (011, 014)
023:          FOR,     (null), (011, 018)
024:   IDENTIFIER, TestStruct, (011, 029)
025:     FUNCTION,     (null), (012, 010)
026:   IDENTIFIER, testFunction, (012, 023)
027:            (,     (null), (012, 024)
028:   IDENTIFIER,       self, (012, 028)
029:            :,     (null), (012, 029)
030:            *,     (null), (012, 031)
031:   IDENTIFIER,       Self, (012, 035)
032:            ),     (null), (012, 036)
033:           ->,     (null), (012, 038)
034:   IDENTIFIER,       Self, (012, 043)
035:           DO,     (null), (012, 046)
036:       RETURN,     (null), (013, 010)
037:            (,     (null), (013, 011)
038:   IDENTIFIER,       self, (013, 015)
039:            ),     (null), (013, 016)
040:            ;,     (null), (013, 017)
041:          END,     (null), (014, 005)
042:          END,     (null), (015, 003)

### AST GENERATION ###
Declaration STRUCT
  Arg IDENTIFIER
    Identifier_Arg x
    Identifier_Arg y
      Type TYPE INT8
Declaration TRAIT TestTrait
  Declaration FUNCTION  testFunction
    Arg IDENTIFIER
      Identifier_Arg self
        Type POINTER
          Type IDENTIFIER Self
    Type IDENTIFIER Self
Declaration IMPL
  Type IDENTIFIER TestTrait
  Type IDENTIFIER TestStruct
  Declaration FUNCTION  testFunction
    Arg IDENTIFIER
      Identifier_Arg self
        Type POINTER
          Type IDENTIFIER Self
    Type IDENTIFIER Self
    Stmt BUILTIN return
      Expr VARIABLE self
